core/network/types/block_announce.hpp

```diff
@@ -7,17 +7,62 @@
 #define KAGOME_BLOCK_ANNOUNCE_HPP
 
 #include "primitives/block_header.hpp"
-#include "scale/tie.hpp"
+#include "scale/scale.hpp"
 
 namespace kagome::network {
-  /**
-   * Announce a new complete block on the network
-   */
-  struct BlockAnnounce {
-    SCALE_TIE(1);
 
+  /// Block state in the chain.
+  enum class BlockState : uint8_t {
+    /// Block is not part of the best chain.
+    Normal,
+    /// Latest best block.
+    Best,
+  };
+
+  /// Announce a new complete relay chain block on the network.
+  struct BlockAnnounce {
+    /// New block header.
     primitives::BlockHeader header;
+
+    /// Block state.
+    std::optional<BlockState> state = std::nullopt;
+
+    /// Data associated with this block announcement, e.g. a candidate message.
+    std::optional<std::vector<uint8_t>> data = std::nullopt;
+
+    friend inline scale::ScaleEncoderStream &operator<<(
+        scale::ScaleEncoderStream &s, const BlockAnnounce &v) {
+      s << v.header;
+      if (v.state.has_value()) {
+        s << v.state.value();
+      }
+      if (v.data.has_value()) {
+        s << v.data.value();
+      }
+      return s;
+    }
+
+    friend inline scale::ScaleDecoderStream &operator>>(
+        scale::ScaleDecoderStream &s, BlockAnnounce &v) {
+      s >> v.header;
+      try {
+        BlockState tmp;
+        s >> tmp;
+        v.state.emplace(tmp);
+      } catch (...) {
```

Comment:
Exceptions make debugger stop, so successful scenario should not throw
```suggestion
      if (s.hasMore(1)) {
        BlockState tmp;
        s >> tmp;
        v.state.emplace(tmp);
      } else {
```

Reply:
Makes sense. Good point. Thanks